### PR TITLE
Prediction dtype fix

### PR DESCRIPTION
**Pull request by bstriner**
Alternative logic to https://github.com/fchollet/keras/pull/5903. Convert to floatx if dtype is float32 or float64, otherwise keep dtype. Only affects predict loop.

Most people are probably using float outputs and are unaffected. If you have int outputs, they will now be predicted as ints.

Cheers

### CODE REVIEW

**Comment by fchollet** [comment]
If which cases would have int outputs?

**Comment by bstriner** [comment]
@fchollet NLP-related tasks are the most likely to have int outputs (at least during test/generation). I have a model with a recurrent unit outputting characters each step and looping that output back into the unit. The choice at each step is discrete so the output is an int, which goes through an embedding at the next step. With most sequence generation models your output will be either an argmax or a sample from a softmax.

I feel like someone could accidentally output the wrong type of float and want it to be automatically converted to the right size, so the existing behavior makes sense. If your output is an int, it is probably intentional, so it shouldn't be converted automatically.

Side note: this behavior isn't ideal. If floatX=float32, and the gpu is returning float64, we are downcasting in numpy. Ideally, we would add a cast in the backend and return only half the data size to the cpu. Maybe a better solution would be to enforce the downcasting at the gpu instead of in numpy, but not really something worth working on right now.

Cheers

**Comment by bstriner** [comment]
Not sure if it is necessary, but a more complicated alternative would be to assume that int size should mirror floatX.
* Convert float32 or float64 to floatX.
* If floatX=float32, convert ints to int32
* If floatX=float64, convert ints to int64
* Leave any other data types unchanged

**Comment by bstriner** [comment]
Just updated to include float16. Float16, 32 and 64 are converted to floatX, everything else is left alone.

If the dtype is bool, int, unsignedint, or complex, that was probably intentional and the type is left alone.

Some really cool stuff out there on complex valued networks so we are future-proof :-)

**Comment by fchollet** [comment]
Are there any reasons why your first PR would not be sufficient (using the data type of the first batch output)?

**Comment by bstriner** [comment]
This one is more similar to current functionality. If there is someone who has float64 outputs and is expecting keras to automatically downcast to float32 they get to keep that feature.

I can imagine the average Joe accidentally having float64 outputs but meaning to have float32, or the other way around.

If you aren't worried about someone who depends on the predict function to downcast or upcast values from the GPU, which they really shouldn't be doing anyways, then the first PR is really the way to go.

As ever, the balance between clean and simple code and maintaining legacy functionality is in your hands.

**Comment by fchollet** [comment]
Let's go with the first PR then.
